api/envoy/api/v2/rds.proto

```diff
@@ -44,7 +45,23 @@ service RouteDiscoveryService {
   }
 }
 
-// [#comment:next free field: 9]
+// Virtual Host Discovery Service (VHDS) is used to dynamically update the list of virtual hosts for
+// a given RouteConfiguration. If VHDS is configured a virtual host list update will be triggerred
+// during the processing of an HTTP request if a route for the request cannot be resolved. The
```

Comment:
Is this actually true? Do we plan on pausing request processing and making a VHDS request? IMO this should probably be done as a separate filter if we want behavior. cc @htuch @dmitri-d

Reply:
@mattklein123 how would this work as an HTTP filter? Don't we need an initial route/cluster pick before we start down the HTTP filter chain? I think we were thinking of augmenting HCM to handle this.

Reply:
No, we still iterate all filters even if there is no route, so we can have an optional filter which pauses and fetches a new VHDS entry if users want that functionality vs. just standard eventually consistent VHDS. I would argue this is a cleaner design.

Reply:
@mattklein123 yeah, it is arguably much cleaner. What to do about things that are done by HCM in `decodeHeaders` before filter execution though, e.g. https://github.com/envoyproxy/envoy/blob/1cf59d4f47ad7ab1a4e924cd66b825bb758f83c7/source/common/http/conn_manager_impl.cc#L701? Ideally, VHDS should be as expressive and capable as regular `RouteConfiguration`.

Reply:
I don't have all the answers right now, but I think this code should be moved into the route refresh code anyway, since arguably the timeout should be changed if the route changes. This would then "just work" on the on-demand VHDS filter. I'm pretty sure we can make this work cleanly as a filter.

Reply:
That sounds good. So, there will need to be some HCM changes, but we'll push most of the logic into a filter. @dmitri-d WDYT?

Reply:
That's pretty close to the WIP implementation (see https://github.com/envoyproxy/envoy/pull/6406/files#diff-be24bc0fe90bba5c4e871fb56928e085R12). An (the main?) issue with moving code into ```refreshCachedRoute``` is probably going to be here: https://github.com/envoyproxy/envoy/blob/1cf59d4f47ad7ab1a4e924cd66b825bb758f83c7/source/common/http/conn_manager_impl.cc#L686. It relies on a route map being present to build a filter chain. I need to think what can be done about that.

Reply:
cc @alyssawilk for comment. I think we might be able to handle this via an internal redirect from the filter if we can't figure out a more elegant solution during route cache reload.